File: backend/tests/fixtures/knowledge_base.py
from __future__ import annotations

from datetime import datetime, timezone

from app.models import (
    KnowledgeGap,
    KnowledgeGapFollow,
    KnowledgeGapNotice,
    KnowledgeSource,
    Textbook,
    TextbookExtensionResource,
    TextbookSectionContent,
)

_PUBLISHED_AT = datetime(2026, 6, 1, tzinfo=timezone.utc)
_ARCHIVED_AT = datetime(2026, 6, 2, tzinfo=timezone.utc)


def enabled_source(
    source_id: str = "source-admitted",
    name: str = "已准入来源",
) -> KnowledgeSource:
    return KnowledgeSource(
        source_id=source_id,
        name=name,
        base_url="https://example.test/source",
        status="enabled",
        source_kind="open_textbook",
        download_requirement="公开下载",
        ai_search_requirement="允许搜索",
        download_status="verified",
        parse_status="supported",
        license_review_status="approved",
        human_review_status="reviewed",
    )


def unenabled_source(
    source_id: str = "source-blocked",
    name: str = "未启用来源",
) -> KnowledgeSource:
    source = enabled_source(source_id=source_id, name=name)
    source.status = "disabled"
    return source


def source_payload(source: KnowledgeSource) -> dict[str, object]:
    return {
        "source_id": source.source_id,
        "name": source.name,
        "base_url": source.base_url,
        "status": source.status,
        "source_kind": source.source_kind,
        "download_requirement": source.download_requirement,
        "ai_search_requirement": source.ai_search_requirement,
        "download_status": source.download_status,
        "parse_status": source.parse_status,
        "license_review_status": source.license_review_status,
        "human_review_status": source.human_review_status,
    }


def admitted_source_payload(
    source_id: str = "source-admitted-api",
) -> dict[str, object]:
    return source_payload(enabled_source(source_id=source_id, name="已准入来源"))


def blocked_source_payload(source_id: str = "source-blocked-api") -> dict[str, object]:
    return source_payload(unenabled_source(source_id=source_id, name="未启用来源"))


def textbook(
    *,
    textbook_id: str,
    title: str,
    source_id: str = "source-admitted",
    description: str = "",
    tags: list[str] | None = None,
    outline: dict[str, object] | None = None,
    ingestion_status: str = "not_started",
    outline_review_status: str = "unreviewed",
    student_availability_status: str = "draft",
    language: str = "en",
    translated_language: str = "zh",
) -> Textbook:
    return Textbook(
        textbook_id=textbook_id,
        source_id=source_id,
        title=title,
        original_title="Linear Algebra",
        language=language,
        translated_language=translated_language,
        description=description,
        tags=tags or [],
        download_url="https://example.test/book.pdf",
        file_asset_url="https://example.test/book.md",
        outline=outline or {},
        ingestion_status=ingestion_status,
        outline_review_status=outline_review_status,
        student_availability_status=student_availability_status,
    )


def published_textbook(
    textbook_id: str = "textbook-published",
    source_id: str = "source-admitted",
    title: str = "线性代数",
) -> Textbook:
    row = textbook(
        textbook_id=textbook_id,
        source_id=source_id,
        title=title,
        description="覆盖矩阵和向量空间",
        tags=["矩阵", "向量空间"],
        outline={"sections": [{"section_id": "1.1", "title": "矩阵乘法"}]},
        ingestion_status="completed",
        outline_review_status="approved",
        student_availability_status="published",
    )
    row.published_at = _PUBLISHED_AT
    return row


def unpublished_textbook(
    textbook_id: str = "textbook-draft",
    source_id: str = "source-admitted",
    title: str = "矩阵专题草稿",
) -> Textbook:
    return textbook(
        textbook_id=textbook_id,
        source_id=source_id,
        title=title,
        description="未发布教材不能进入上下文",
        tags=["矩阵"],
        outline={"sections": [{"section_id": "1.1", "title": "矩阵"}]},
        ingestion_status="completed",
        outline_review_status="approved",
        student_availability_status="draft",
    )


def archived_textbook(
    textbook_id: str = "textbook-archived",
    source_id: str = "source-admitted",
    title: str = "已归档教材",
) -> Textbook:
    row = unpublished_textbook(
        textbook_id=textbook_id,
        source_id=source_id,
        title=title,
    )
    row.student_availability_status = "archived"
    row.archived_at = _ARCHIVED_AT
    return row


def textbook_payload(row: Textbook) -> dict[str, object]:
    return {
        "textbook_id": row.textbook_id,
        "source_id": row.source_id,
        "title": row.title,
        "original_title": row.original_title,
        "language": row.language,
        "translated_language": row.translated_language,
        "description": row.description,
        "tags": row.tags,
        "download_url": row.download_url,
        "file_asset_url": row.file_asset_url,
        "outline": row.outline,
        "ingestion_status": row.ingestion_status,
        "outline_review_status": row.outline_review_status,
        "student_availability_status": row.student_availability_status,
        "ingestion_error_message": row.ingestion_error_message,
    }


def section(
    *,
    textbook_id: str,
    section_content_id: str,
    section_id: str,
    title: str = "小节",
    content_zh: str,
    content_original: str = "",
    order_index: int = 1,
    parent_section_id: str | None = None,
    original_title: str = "",
    content_char_count: int | None = None,
) -> TextbookSectionContent:
    return TextbookSectionContent(
        section_content_id=section_content_id,
        textbook_id=textbook_id,
        section_id=section_id,
        parent_section_id=parent_section_id,
        order_index=order_index,
        title=title,
        original_title=original_title,
        content_original=content_original or content_zh,
        content_zh=content_zh,
        content_char_count=(
            len(content_zh) if content_char_count is None else content_char_count
        ),
    )


def section_payload(row: TextbookSectionContent) -> dict[str, object]:
    return {
        "section_content_id": row.section_content_id,
        "section_id": row.section_id,
        "parent_section_id": row.parent_section_id,
        "order_index": row.order_index,
        "title": row.title,
        "original_title": row.original_title,
        "content_original": row.content_original,
        "content_zh": row.content_zh,
        "content_char_count": row.content_char_count,
    }


def continuous_sections(
    textbook_id: str = "textbook-evidence",
) -> list[TextbookSectionContent]:
    return [
        section(
            textbook_id=textbook_id,
            section_content_id=f"section-evidence-{index}",
            section_id=f"1.{index}",
            order_index=index,
            title=f"小节 {index}",
            content_zh=f"第 {index} 节中文正文",
        )
        for index in range(1, 4)
    ]


def non_continuous_sections(
    textbook_id: str = "textbook-evidence",
) -> list[TextbookSectionContent]:
    return [
        section(
            textbook_id=textbook_id,
            section_content_id="section-evidence-non-continuous-1",
            section_id="1.1",
            order_index=1,
            title="小节 1",
            content_zh="第 1 节中文正文",
        ),
        section(
            textbook_id=textbook_id,
            section_content_id="section-evidence-non-continuous-3",
            section_id="1.3",
            order_index=3,
            title="小节 3",
            content_zh="第 3 节中文正文",
        ),
    ]


def over_8000_char_sections(
    textbook_id: str = "textbook-evidence",
) -> list[TextbookSectionContent]:
    return [
        section(
            textbook_id=textbook_id,
            section_content_id="section-evidence-over-limit-1",
            section_id="1.98",
            order_index=98,
            title="超长小节一",
            content_zh="字" * 4001,
        ),
        section(
            textbook_id=textbook_id,
            section_content_id="section-evidence-over-limit-2",
            section_id="1.99",
            order_index=99,
            title="超长小节二",
            content_zh="字" * 4001,
        ),
    ]


def structured_textbook_payload(
    textbook_id: str = "textbook-linear-api",
    source_id: str = "source-admitted-api",
    title: str = "线性代数",
    outline_review_status: str = "approved",
) -> dict[str, object]:
    row = unpublished_textbook(
        textbook_id=textbook_id,
        source_id=source_id,
        title=title,
    )
    row.description = "覆盖矩阵乘法"
    row.tags = ["矩阵"]
    row.outline = {"sections": [{"section_id": "1.1", "title": "矩阵乘法"}]}
    row.outline_review_status = outline_review_status
    rows = [
        section(
            textbook_id=textbook_id,
            section_content_id=f"section-{textbook_id}-1",
            section_id="1.1",
            order_index=1,
            title="矩阵乘法",
            original_title="Matrix Multiplication",
            content_zh="矩阵乘法的中文正文。",
            content_char_count=0,
        )
    ]
    return {
        "textbook": textbook_payload(row),
        "sections": [section_payload(section_row) for section_row in rows],
    }


def uncovered_topic_gap(
    gap_id: str = "gap-api",
    normalized_topic: str = "概率论",
) -> KnowledgeGap:
    return KnowledgeGap(
        gap_id=gap_id,
        normalized_topic=normalized_topic,
        trigger_count=1,
        status="open",
        student_goal_summaries=["希望学习概率基础"],
    )


def followed_gap_student(
    gap_id: str = "gap-api",
    user_uid: str = "user-followed-gap",
    follow_id: str = "gap-follow-followed-student",
) -> KnowledgeGapFollow:
    return KnowledgeGapFollow(
        follow_id=follow_id,
        gap_id=gap_id,
        user_uid=user_uid,
    )


def gap_resolved_notice(
    *,
    notice_id: str,
    gap_id: str,
    user_uid: str,
    learning_topic: str = "概率论",
    textbook_id: str = "textbook-linear-api",
) -> KnowledgeGapNotice:
    return KnowledgeGapNotice(
        notice_id=notice_id,
        gap_id=gap_id,
        user_uid=user_uid,
        notice_type="knowledge_gap_resolved",
        title=f"{learning_topic} 已补齐",
        body="知识库已发布覆盖该主题的教材。",
        action_label="重新生成学习路径",
        action_payload={
            "action": "regenerate_learning_path_intake",
            "learning_topic": learning_topic,
            "textbook_id": textbook_id,
        },
    )


def extension_resource(
    *,
    resource_id: str,
    textbook_id: str,
    section_id: str,
    title_zh: str,
    status: str,
    render_mode: str = "webpage",
) -> TextbookExtensionResource:
    return TextbookExtensionResource(
        resource_id=resource_id,
        textbook_id=textbook_id,
        section_id=section_id,
        resource_type="webpage",
        title_zh=title_zh,
        description_zh="补充阅读",
        render_mode=render_mode,
        url=f"https://example.test/{resource_id}",
        cover_url=f"https://example.test/{resource_id}.png",
        source_name="已准入来源",
        status=status,
    )


def extension_resource_payload(row: TextbookExtensionResource) -> dict[str, object]:
    return {
        "resource_id": row.resource_id,
        "section_id": row.section_id,
        "resource_type": row.resource_type,
        "title_zh": row.title_zh,
        "description_zh": row.description_zh,
        "render_mode": row.render_mode,
        "url": row.url,
        "cover_url": row.cover_url,
        "source_name": row.source_name,
        "status": row.status,
    }


def extension_resources_three(
    textbook_id: str = "textbook-extension",
    section_id: str = "1.1",
) -> list[TextbookExtensionResource]:
    return [
        extension_resource(
            resource_id=f"resource-visible-{index}",
            textbook_id=textbook_id,
            section_id=section_id,
            title_zh=f"扩展资料 {index}",
            status="published",
        )
        for index in range(1, 4)
    ]


def extension_resources_four(
    textbook_id: str = "textbook-extension",
    section_id: str = "1.1",
) -> list[TextbookExtensionResource]:
    return [
        extension_resource(
            resource_id=f"resource-visible-{index}",
            textbook_id=textbook_id,
            section_id=section_id,
            title_zh=f"扩展资料 {index}",
            status="published",
        )
        for index in range(1, 5)
    ]

File: frontend/src/test/fixtures/knowledgeBase.ts
import type { components } from "../../types/api";
import type { ChatMessage, SessionMessage } from "../../types/chat";

type KnowledgeGapRead = components["schemas"]["KnowledgeGapRead"];
type KnowledgeGapFollowRead = components["schemas"]["KnowledgeGapFollowRead"];
type KnowledgeGapNoticeRead = components["schemas"]["KnowledgeGapNoticeRead"];
type KnowledgeSourceRead = components["schemas"]["KnowledgeSourceRead"];
type TextbookExtensionResourceRead =
	components["schemas"]["TextbookExtensionResourceRead"];
type TextbookRead = components["schemas"]["TextbookRead"];
type TextbookSectionContentFixture =
	components["schemas"]["TextbookSectionContentCreateRequest"] & {
		textbook_id: string;
	};

const createdAt = "2026-06-27T10:00:00Z";

export const enabledKnowledgeSource: KnowledgeSourceRead = {
	source_id: "source-enabled-api",
	name: "开放教材来源",
	base_url: "https://example.edu/open",
	status: "enabled",
	source_kind: "open_textbook",
	download_requirement: "direct",
	ai_search_requirement: "allowed",
	download_status: "verified",
	parse_status: "supported",
	license_review_status: "approved",
	human_review_status: "reviewed",
};

export const unpublishedKnowledgeTextbook: TextbookRead = {
	textbook_id: "textbook-draft-api",
	source_id: enabledKnowledgeSource.source_id,
	title: "概率论草稿",
	original_title: "Probability Draft",
	language: "en",
	translated_language: "zh",
	description: "覆盖概率论基础，尚未发布。",
	tags: ["概率论"],
	download_url: "https://example.edu/probability.pdf",
	file_asset_url: "/assets/probability.pdf",
	outline: { sections: [{ section_id: "1", title: "随机事件" }] },
	ingestion_status: "ready_for_outline_review",
	outline_review_status: "unreviewed",
	student_availability_status: "draft",
	ingestion_error_message: "",
	published_at: null,
	unpublished_at: null,
	archived_at: null,
};

export const publishedKnowledgeTextbook: TextbookRead = {
	...unpublishedKnowledgeTextbook,
	textbook_id: "textbook-linear-api",
	title: "线性代数",
	description: "覆盖矩阵、向量空间和概率论衔接。",
	tags: ["矩阵", "概率论"],
	ingestion_status: "completed",
	outline_review_status: "approved",
	student_availability_status: "published",
	published_at: createdAt,
};

export const archivedKnowledgeTextbook: TextbookRead = {
	...publishedKnowledgeTextbook,
	textbook_id: "textbook-archived-api",
	title: "已归档教材",
	student_availability_status: "archived",
	archived_at: "2026-06-28T10:00:00Z",
};

export const continuousKnowledgeSections: TextbookSectionContentFixture[] = [
	{
		section_content_id: "section-content-1",
		textbook_id: publishedKnowledgeTextbook.textbook_id,
		section_id: "1.1",
		parent_section_id: "1",
		order_index: 1,
		title: "矩阵乘法",
		original_title: "Matrix Multiplication",
		content_original: "Matrix multiplication original content.",
		content_zh: "矩阵乘法正文。",
		content_char_count: 7,
	},
	{
		section_content_id: "section-content-2",
		textbook_id: publishedKnowledgeTextbook.textbook_id,
		section_id: "1.2",
		parent_section_id: "1",
		order_index: 2,
		title: "向量空间",
		original_title: "Vector Spaces",
		content_original: "Vector spaces original content.",
		content_zh: "向量空间正文。",
		content_char_count: 7,
	},
];

export const overLimitKnowledgeSections: TextbookSectionContentFixture[] = [
	{
		...continuousKnowledgeSections[0],
		section_content_id: "section-content-over-limit-1",
		section_id: "2.1",
		order_index: 10,
		content_zh: "概".repeat(4100),
		content_char_count: 4100,
	},
	{
		...continuousKnowledgeSections[1],
		section_content_id: "section-content-over-limit-2",
		section_id: "2.2",
		order_index: 11,
		content_zh: "率".repeat(4100),
		content_char_count: 4100,
	},
];

export const uncoveredKnowledgeGap: KnowledgeGapRead = {
	gap_id: "gap-api",
	normalized_topic: "概率论",
	trigger_count: 1,
	follow_count: 0,
	latest_triggered_at: createdAt,
	student_goal_summaries: ["希望学习概率基础"],
	status: "open",
	resolved_textbook_id: null,
	resolved_at: null,
};

export const followedKnowledgeGap: KnowledgeGapFollowRead = {
	follow_id: "gap-follow-followed-student",
	gap_id: uncoveredKnowledgeGap.gap_id,
	user_uid: "user-followed-gap",
	created_at: createdAt,
};

export const resolvedKnowledgeGapNotice: KnowledgeGapNoticeRead = {
	notice_id: "gap-notice-api",
	gap_id: uncoveredKnowledgeGap.gap_id,
	user_uid: "user-1",
	notice_type: "knowledge_gap_resolved",
	title: "概率论 已补齐",
	body: "知识库已发布覆盖该主题的教材。",
	action_label: "重新生成学习路径",
	action_payload: {
		action: "regenerate_learning_path_intake",
		learning_topic: uncoveredKnowledgeGap.normalized_topic,
		textbook_id: publishedKnowledgeTextbook.textbook_id,
	},
	read_at: null,
	created_at: "2026-06-28T10:00:00Z",
};

export const extensionResourcesThree: TextbookExtensionResourceRead[] = [
	"reader",
	"video",
	"webpage",
].map((renderMode, index) => ({
	resource_id: `resource-${index + 1}`,
	textbook_id: publishedKnowledgeTextbook.textbook_id,
	section_id: continuousKnowledgeSections[0].section_id,
	resource_type: "supplement",
	title_zh: `扩展资料 ${index + 1}`,
	description_zh: "用于补充当前章节。",
	render_mode: renderMode as TextbookExtensionResourceRead["render_mode"],
	url: `https://example.edu/resource-${index + 1}`,
	cover_url: "",
	source_name: "开放教材来源",
	status: "active",
}));

export const extensionResourcesFour: TextbookExtensionResourceRead[] = [
	...extensionResourcesThree,
	{
		...extensionResourcesThree[0],
		resource_id: "resource-4",
		title_zh: "扩展资料 4",
		url: "https://example.edu/resource-4",
	},
];

export const profileWithKnowledgeGap: SessionMessage = {
	type: "collecting",
	stage: "goal_constraint",
	question_mode: "none",
	confirmed_info: {
		current_grade: "大三",
		major: "软件工程",
		learning_stage: "有基础",
		has_clear_goal: "大致有方向",
		learning_method_preference: "项目驱动学习",
		learning_pace_preference: "按项目里程碑推进",
		content_preference: ["代码实践"],
		need_guidance: "需要轻量提醒",
		knowledge_foundation: "软件工程基础",
		strengths: "工程实现",
		weaknesses: "概率基础薄弱",
		experience: "做过课程项目",
		short_term_goal: "补齐概率论",
		long_term_goal: "形成 AI 应用开发能力",
		weekly_available_time: "每周 8 小时",
		constraints: "时间有限",
	},
	defaulted_fields: [],
	question_md: "当前知识库还没有覆盖概率论。",
	question_box: {
		question: "",
		options: [],
	},
	text: "当前知识库还没有覆盖概率论。你可以关注该主题，补齐后会收到提醒。",
	gap_id: uncoveredKnowledgeGap.gap_id,
};

export const knowledgeGapProfileSessionCache = {
	userUid: "user-1",
	savedAt: 1000,
	messages: [
		{
			id: "knowledge-gap-user-message",
			role: "user",
			content: "我想学概率论",
			status: "completed",
			timestamp: 1000,
		},
		{
			id: "knowledge-gap-assistant-message",
			role: "assistant",
			content: profileWithKnowledgeGap.text,
			status: "completed",
			timestamp: 1001,
			sessionMessage: {
				...profileWithKnowledgeGap,
				type: "collecting",
			},
			runTrace: [],
			activeStepId: null,
		},
	],
} satisfies {
	userUid: string;
	savedAt: number;
	messages: ChatMessage[];
};
